File: tvm_demo_cpu.py
from tvm.contrib import graph_executor
from tvm.driver.tvmc.model import TVMCPackage
from tvm.driver.tvmc.runner import get_input_info, make_inputs_dict
from tvm.driver import tvmc
import tvm.runtime
from tvm import relay
from tvm import rpc
import tvm
import time
from PIL import Image
import numpy as np
import cv2




# 准备模型
ctx = tvm.cpu(0)
tvmc_package = TVMCPackage(package_path="resnet18-cpu.tar")
lib = tvm.runtime.load_module(tvmc_package.lib_path)
m = graph_executor.create(tvmc_package.graph, lib, ctx)
m.load_params(tvmc_package.params)
print("model loaded")


# 准备数据
mean = [0.485, 0.456, 0.406]
std = [0.229, 0.224, 0.225]
image = cv2.imread("cat.jpg")
image = cv2.resize(image, (224, 224))
image = cv2.cvtColor(image, cv2.COLOR_BGR2RGB)
image = image / 255.0
image = image - mean
image = image / std
image = image.transpose((2, 0, 1))
image_batch = np.expand_dims(image, axis=0).astype(np.float32)
image_batch = np.repeat(image_batch, 32, axis=0)
image_batch_tvm = tvm.nd.array(image_batch, ctx)


# 推理
shape_dict, dtype_dict = get_input_info(tvmc_package.graph, tvmc_package.params)
print(f"shape_dict:{shape_dict}")
print(f"dtype_dict:{dtype_dict}")
m.set_input("input", image_batch_tvm)
m.run()

# 后处理
output = m.get_output(0).asnumpy()
class_cat = output[0]
softmax_cat = np.exp(class_cat) / np.sum(np.exp(class_cat))
index_of_cat = np.argmax(softmax_cat)
print("285 n02124075 猫, Egyptian cat")
print(f"cat.jpg: {index_of_cat}")
print(f"cat.jpg score:{softmax_cat[index_of_cat]}")


# 耗时测试
t0 = time.time()
for i in range(100):
    m.set_input("input", image_batch_tvm)
    m.run()
c = time.time() - t0
print(f"infer 100 cost average: {c/100}")


# output
# model loaded
# shape_dict:{'input': [32, 3, 224, 224]}
# dtype_dict:{'input': 'float32'}
# 285 n02124075 猫, Egyptian cat
# cat.jpg: 285
# cat.jpg score:0.4857271611690521
# infer 100 cost average: 0.38649285316467286 

File: tvm_demo_cpu.cpp
#include <dlpack/dlpack.h>
#include <tvm/runtime/module.h>
#include <tvm/runtime/packed_func.h>
#include <tvm/runtime/registry.h>
#include <opencv2/opencv.hpp>
#include <cstdio>
#include <vector>

using namespace std;


bool pairCompare(const std::pair<float, int>& lhs, const std::pair<float, int>& rhs)
{
  return lhs.first > rhs.first;
}

std::vector<int> argmax(const std::vector<float>& v, int topK)
{
	std::vector<std::pair<float, int>> pairs(v.size());
	for (size_t i = 0; i < v.size(); ++i)
		pairs[i] = std::make_pair(v[i], i);

	std::partial_sort(pairs.begin(), pairs.begin() + topK, pairs.end(), pairCompare);

	std::vector<int> result;
	result.reserve(topK);
	for (int i = 0; i < topK; ++i)
		result.push_back(pairs[i].second);
	return result;
}

int main(void) {
  // load the ResNet18 library
  DLDevice dev{kDLCPU, 0};
  tvm::runtime::Module mod_factory = tvm::runtime::Module::LoadFromFile("mod.so");
  
  // create the ResNet18 module
  tvm::runtime::Module resnet18_mod = mod_factory.GetFunction("default")(dev);
  tvm::runtime::PackedFunc set_input = resnet18_mod.GetFunction("set_input");
  tvm::runtime::PackedFunc get_output = resnet18_mod.GetFunction("get_output");
  tvm::runtime::PackedFunc run = resnet18_mod.GetFunction("run");

  // Use the C++ API
  // Replace the input size and data type according to your ResNet18 model
  tvm::runtime::NDArray input = tvm::runtime::NDArray::Empty({32, 3, 224, 224}, DLDataType{kDLFloat, 32, 1}, dev);
  tvm::runtime::NDArray output = tvm::runtime::NDArray::Empty({32, 1000}, DLDataType{kDLFloat, 32, 1}, dev);

  // Set input data (replace with your input data preparation logic)
  // read image
  cv::Mat image = cv::imread("cat.jpg");
  cv::resize(image, image, cv::Size(224, 224));
  cv::cvtColor(image, image, cv::COLOR_BGR2RGB);
  vector<float> mean{0.485, 0.456, 0.406}, std{0.229, 0.224, 0.225};
  // copy
  float* input_data = static_cast<float*>(input->data);
  std::vector<cv::Mat> channels(3);
  cv::split(image, channels);
  for(int b = 0; b<32;b++){
    for(int i=0;i<3;i++){
      for(int j=0;j<224;j++){
        for(int k=0;k<224;k++){
          float point = static_cast<float>(channels[i].at<uchar>(j, k));
          input_data[b*3*224*224 + i*224*224 + j*224 + k] = (point / 255.0 - mean[i]) / std[i];
        }
      }
    }
  }

  // Set the input
  set_input("input", input);
  
  // Run the ResNet18 model
  run();

  // Get the output
  get_output(0, output);

  // 打印Tips
  printf("******Tips:\n");
  string tips = string("280 n02120505 狐狸, grey fox, gray fox, Urocyon cinereoargenteus\n")+
                "281 n02123045 猫, tabby, tabby cat\n"+
                "282 n02123159 猫, tiger cat\n"+
                "283 n02123394 猫, Persian cat\n"+
                "284 n02123597 猫, Siamese cat, Siamese\n"+
                "285 n02124075 猫, Egyptian cat\n"+
                "286 n02125311 猫, cougar, puma, catamount, mountain lion, painter, panther, Felis concolor\n"+
                "287 n02127052 猫, lynx, catamount\n"+
                "288 n02128385 豹, leopard, Panthera pardus\n"+
                "289 n02128757 豹, snow leopard, ounce, Panthera uncia\n"+
                "290 n02128925 豹, jaguar, panther, Panthera onca, Felis onca\n";
  std::cout<<tips;
  // Add your post-processing logic here
  float* output_data = static_cast<float*>(output->data);
  for(int b = 0; b<32; b++){
    float sum = 0;
    std::vector<float> probs(1000);

    // 累加
    for(int idx=0;idx<1000;idx++){
      probs[idx] = exp(output_data[b * 1000 + idx]);
      sum += probs[idx];
    }
    for(int idx=0;idx<1000;idx++){
      probs[idx] /= sum;
    }

    // 找出最大的前5个
    std::vector<int> topK = argmax(probs, 5);
    if(b == 0 || b==1){
      printf("******Top 5:\n");
      for(int idx=0;idx<5;idx++){
        printf("Batch %d: %d, %f\n", b, topK[idx], probs[topK[idx]]);
      }
    }

  }

  return 0;
}